### ISSUE TITLE
Type inference doesn't appear to be working on a generic

### ISSUE THREAD

**Issue by DataTables** [Microsoft/TypeScript#19376]
**TypeScript Version:**  2.5.3

**Code**

```ts
interface SocketHandler {
    <T>(
        locals: string,
        payload: T,
        callback: Function
    ): void;
}

interface SocketRouter {
    (
        eventName: string,
        handler: SocketHandler
    ): void;
}

interface StaffJSON {
    count: number;
    record: string;
}

class Staff {
    public count (): number {
        return 3; // would do some db interaction
    }
}

let router: SocketRouter = function ( name, handler ) {
    // Save the name and handler
}

router( 'staff-count', function<StaffJSON>(locals, payload, cb) {
    payload.record = 1;
    let staff = new Staff();
    cb( staff.count() );
} );
```

**Expected behavior:**

Expected this to result in an error since `record` is marked as a string, but a number is assigned.

**Actual behavior:**

The compiler doesn't complain at all. Moreover, the type inference displayed in VS Code shows that "payload" (and the other arguments of the `router` call's callback function) are of type "any". I expected the Generic to set the `payload` parameter to be `StaffJSON` type, while `locals` would be a string, etc.

My goal is to have `router` exposed from a module which can be included wherever requires - only the `payload` needs to be defined as a specific type (e.g. a different JSON object) for each call. Its somewhat similar to an Express router, hence my desire to keep `router` generic.

I'm guessing that this is a limitation in the type inference of the generics, and while I've read various other issues on this, I'm not really any closer to wrapping my head around it I'm afraid. My apologies if it has come up before and I haven't found it.

I presume there is a way to get this to work somehow - being able to type check what I'm using the `payload` for is why TypeScript is so attractive.

**Comment by andy-ms** [comment]
Try enabling `--noImplicitAny` and you'll get an error.
The problem is that `function<StaffJSON>(locals, payload, cb)` doesn't do what you think: it declares a function with a type *parameter*, and then doesn't use that type parameter. Since the function has a type parameter it doesn't get a contextual type and all of its parameters  are `any`.
Remove `<StaffJSON>` and you'll see that the types are `locals: string, payload: T, cb: Function`.

The problem is that `SocketRouter` declares to take a `SocketHandler`, and `SocketHandler` is a type of *generic* functions, which can only be satisfied by another generic function.
If you intended that a `SocketHandler` implementation can substitute anything for `T`, you might as well do away with the generics and just declare `payload: any`.

**Comment by jcalz** [comment]
This feels like it should be [a Stack Overflow question](https://stackoverflow.com/questions/tagged/typescript) to me.   Maybe this is what you want to use, instead:

```ts
type SocketHandler<T> = (locals: string, payload: T, callback: Function) => void;

type SocketRouter<T> = (eventName: string, handler: SocketHandler<T>) => void;

interface StaffJSON {
  count: number;
  record: string;
}

declare const router: SocketRouter<StaffJSON>;

router('staff-count', function(locals, payload, cb) {
  payload.record = 1; // error as expected
});
```

If not, or you need more help, consider asking on Stack Overflow.  Good luck.

**Comment by DataTables** [comment]
Thank you both for your replies - I really appreciate it! @andy-ms You are right, it isn't working the way I had anticipated - I thought since I had told it that the second parameter to a function with `SocketRouter` would be of type `SocketHandler`, it would match that signature and infer automatically. You've clarified that point for me - thank you :-).
